### PR TITLE
refactor: start datanode more flexibly

### PR DESCRIPTION
**Pull request by MichaelScofield**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

.. to support our dual read write mode.

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.

### CODE REVIEW

File: src/datanode/src/datanode.rs [hunk 1]
```diff
@@ -171,76 +186,63 @@ impl DatanodeBuilder {
...
-                let meta_client = meta_client.context(MissingMetaClientSnafu)?;
+        self.initialize_region_server(&region_server, kv_backend, !controlled_by_metasrv)
+            .await?;
 
-                let heartbeat_task =
-                    HeartbeatTask::try_new(&self.opts, region_server.clone(), meta_client).await?;
-                Some(heartbeat_task)
-            }
-            Mode::Standalone => None,
+        let heartbeat_task = if let Some(meta_client) = meta_client {
```
Comment (WenyXu) [inline src/datanode/src/datanode.rs:239]:
```suggestion
        let heartbeat_task = if controlled_by_metasrv {
```

File: src/datanode/src/region_server.rs [hunk 2]
```diff
@@ -79,12 +79,27 @@ impl RegionServer {
...
+            event_listener,
+            Arc::new(DummyTableProviderFactory),
+        )
+    }
+
+    pub fn with_table_provider(
+        query_engine: QueryEngineRef,
+        runtime: Arc<Runtime>,
+        event_listener: RegionServerEventListenerRef,
+        table_provider_factory: Arc<dyn TableProviderFactory>,
```
Comment (WenyXu) [inline src/datanode/src/region_server.rs:95]: Nit: prefer to use `TableProviderFactoryRef`

File: src/frontend/src/instance/standalone.rs [hunk 3]
```diff
@@ -107,7 +107,7 @@ impl Datanode for RegionInvoker {
     }
 }
 
-pub(crate) struct StandaloneTableMetadataCreator {
+pub struct StandaloneTableMetadataCreator {
```
Comment (WenyXu) [inline src/frontend/src/instance/standalone.rs:110]: Nit: Adding some tests is better if you use this in the outside crate. The upstream changes might break the downstream code unintentionally. Or log it in the issues and leave it for future work.

**Review by WenyXu — APPROVED** [review]
Rest LGTM

**Review by fengjiachun — APPROVED** [review]

**Pull request merged**